### ISSUE TITLE
Use JetBrains's Java annotations?

### ISSUE THREAD

**Issue by trustin** [line/armeria#2793]
https://github.com/JetBrains/java-annotations/

Annotations like `@Contract` and `@Range` look useful, especially for those who uses IntelliJ IDEA.

**Comment by minwoox** [comment]
Seems like good annotations. I want it. 👍

**Comment by trustin** [comment]
We could also replace the annotations from `javax.annotation` using it, i.e. less issues with Java module system.

**Comment by minwoox** [comment]
> We could also replace the annotations from javax.annotation using it

I found a few more annotations that we can't replace such as `NotThreadSafe`, `GuardedBy` and `CheckReturnValue`.
I think we can just remove `NotThreadSafe` and `GuardedBy`. However, `CheckReturnValue` is useful especially when a user is using errorprone.
So, I think we cannot just replace `javax.annotation`. 🤔

**Comment by anuraaga** [comment]
By the way, everyone has their own annotations package it seems. Not sure if there are any JMS issues still with javax.annotation, but if there are it'd be an easy rename to use a different one.

https://github.com/google/error-prone/tree/master/annotations/src/main/java/com/google/errorprone/annotations

**Comment by trustin** [comment]
https://github.com/jspecify/jspecify is promising but I guess it will take long time until it's ready.

Should we use both `jetbrains-annotations` and `errorprone-annotations`? 🤔

**Comment by trustin** [comment]
> Not sure if there are any JMS issues still with javax.annotation

https://github.com/google/guava/issues/2960

**Comment by anuraaga** [comment]
Yeah huge thread :) Not sure what this means but I've seen mixed messages about whether there's actually a problem anymore in various repos https://github.com/google/guava/issues/2960#issuecomment-431599487

**Comment by minwoox** [comment]
Oh, I didn't know that errorprone just checks the simple name of the annotations. https://github.com/google/error-prone/blob/master/core/src/main/java/com/google/errorprone/bugpatterns/CheckReturnValue.java#L116

**Comment by ikhoon** [comment]
JFYI, JetBrain's annotation would help to generate code for devs.
<img width="757" alt="image" src="https://user-images.githubusercontent.com/1866157/86435644-69903500-bd3b-11ea-87b1-3de7c458a462.png">

**Comment by trustin** [comment]
@ikhoon The warning (?) goes away after clicking 'Make ... default annotation', though.

**Comment by ikhoon** [comment]
> @ikhoon The warning (?) goes away after clicking 'Make ... default annotation', though.

Ah... Please forget the comment. 😅 I misunderstand something.

**Comment by ikhoon** [comment]
> @ikhoon The warning (?) goes away after clicking 'Make ... default annotation', though.

Actually what I wanted to say is "code generation" in IntelliJ 😆

![Kapture 2020-07-03 at 15 45 16](https://user-images.githubusercontent.com/1866157/86439828-41f19a80-bd44-11ea-8147-98a8964d474f.gif)

**Comment by trustin** [comment]
Hmm, I think I only experienced such issue before I made `javax.annotation.Nullable` default. Perhaps a user has to restart the IDE or re-open the editor? 🤔

**Comment by minwoox** [comment]
I realized that the `RetentionPolicy` of the `Nullbale` annotation from Jetbrains is `CLASS` which means that we cannot use it in the annotated service. https://armeria.dev/docs/server-annotated-service#getting-a-query-parameter
And they don't want to change the policy. https://github.com/JetBrains/java-annotations/pull/1

We cannot define our own `Nullable` annotation because it needs `TypeQualifierNickname`, which is in JSR-305, for Kotlin.

Let me revisit this later if we have better options.

**Comment by tobias-** [comment]
> I realized that the `RetentionPolicy` of the `Nullable` annotation from Jetbrains is `CLASS` which means that we cannot use it in the annotated service. https://armeria.dev/docs/server-annotated-service#getting-a-query-parameter
> And they don't want to change the policy. [JetBrains/java-annotations#1](https://github.com/JetBrains/java-annotations/pull/1)
> 
FWIW, this is what got us to actually add both annotations to protobuf generated classes (programmatically). IntelliJ is a little bit flaky when it comes to javax.annotations, and jetbrains annotations wasn't available runtime so we ran into other problems with that (can't remember exactly what).

Imho there is any compelling reason to change, as while some things will improve, some will be worse.

**Comment by ikhoon** [comment]
After `ElementType.TYPE_USE` was applied to `@NonNullByDefault` #3558,
Kotlin disallows to return `null` by a `BiFunction<A,B,C>` by default.
![image](https://user-images.githubusercontent.com/1866157/128114710-8a9ba1ec-10a6-42a0-bfd8-4df8f55563cf.png)
Slack thread: https://line-armeria.slack.com/archives/C1NGPBUH2/p1628020149026800

Unfortunately, we can not apply `javax.annotation.Nullable` to a type parameter of a generic type.
When `@org.jetbrains.annotations.Nullable` was annotated to the return type, the compiler error was gone.
For example:
```java
public LoggingDecoratorBuilder responseCauseSanitizer(
        BiFunction<? super RequestContext, ? super Throwable, @org.jetbrains.annotations.Nullable ?> responseCauseSanitizer) {
    this.responseCauseSanitizer = requireNonNull(responseCauseSanitizer, "responseCauseSanitizer");
    return this;
}
```

> I realized that the RetentionPolicy of the Nullable annotation from Jetbrains is CLASS which means that we cannot use it in the annotated service.

Perhaps, we keep using `javax.annotation.Nullable` for annotated service and Use JetBrains java-annotation for static analysis tools.

**Comment by trustin** [comment]
@ikhoon Sounds good to me. We might want to introduce a new annotation that hides the direct reference to JetBrains' `@Nullable` annotation, though:

```java
@org.jetbrains.annotations.Nullable
...
public @interface NullableType {}

// in LoggingDecoratorBuilder:
... BiFunction<? super RequestContext, ? super Throwable, @NullableType ?> 
```

**Comment by ikhoon** [comment]
I was mistaken for '@org.jetbrains.annotations.Nullable' and wildcard(`?`) types to work because of the compilation cache. 
![image](https://user-images.githubusercontent.com/1866157/128300596-30302659-31f8-4c72-b96c-d4ccbd9e00ca.png)
`Nullable` annotation is not applicable to wildcard types and compilation also fails.

Additionally, AFAIK, JetBrains Java annotations do not work with `@TypeQualifierDefault` and there is no alternative.
That means we can use `org.jetbrains.annotations.NotNull` on the package level.
The following suggestion might make sense to resolve this problem at the moment. Otherwise, we will need to revert #3558
- Keep `javax.annotation` as it is.
- Add `@NullableType` that extends `javax.annotation.Nullable` and enables `ElementType.TYPE_USE`.
  ```java
  @TypeQualifierNickname
  @Nullable
  @Target({ ElementType.METHOD, ElementType.FIELD, ElementType.PARAMETER, ElementType.LOCAL_VARIABLE, ElementType.TYPE_USE })
  public @interface NullableType {}
  ```
- Add new methods that return `Object` or `Void` type instead of wildcard types and deprecate the original methods.
  ```java
  public LoggingDecoratorBuilder responseCauseSanitizerFunction(
        BiFunction<? super RequestContext, ? super Throwable, @NulableType Object> responseCauseSanitizer) {
      ..
  }

  // Will be removed in Armeria 2.0
  @Deprecate
  public LoggingDecoratorBuilder responseCauseSanitizer(
        BiFunction<? super RequestContext, ? super Throwable, ?> responseCauseSanitizer) {
      ...
  }
  ```

**Comment by anuraaga** [comment]
@ikhoon Did you check if you can use checkerframework's Nullable? I think it should support type use.

**Comment by ikhoon** [comment]
> @ikhoon Did you check if you can use checkerframework's Nullable? I think it should support type use. It's what I use in new projects now.

I haven't tried it yet. That sounds nice. Let me check whether it works with Kotlin and IntelliJ static analysis.

**Comment by ikhoon** [comment]
Kotlin understands checkerframework's `Nullable`, however, Kotlin compiler does not produce an error message.
It only helps remove platform-types. https://kotlinlang.org/docs/java-interop.html#notation-for-platform-types
JSR-305 is the only annotation that can produce a compilation level error message with `-Xjsr305=strict`.
https://kotlinlang.org/docs/java-interop.html#compiler-configuration

For better inter-ops with Kotlin, JSR-305 seems the best option at the moment.

**Comment by tobias-** [comment]
TL;DR: The `responseCauseSanitizer { _, _ -> null }` chunk above doesn't build in Gradle anymore since 1.10.0.

It used to be that IntelliJ error'ed, but as long as it was set to build with Gradle, it would still build and be able to run. Now it cannot build the code even in CI, turning it from nuisance into something worse.
